sw/otbn/code-snippets/rsa_verify.s

```diff
@@ -1081,96 +1089,67 @@ modexp_65537:
   lw        x21, 28(x0)
-
-  /* zeroize w2 and reset flags */
-  bn.sub    w2, w2, w2
-
-  /* this loop initializes the output buffer with -M */
-  loop      x30, 3
-    /* load limb from modulus */
-    bn.lid    x11, 0(x16++)
-
-    /* subtract limb from 0 */
-    bn.subb   w2, w31, w2
-
-    /* store limb in dmem */
-    bn.sid    x11, 0(x21++)
-
-  /* reload pointer to 1st limb of modulus */
-  lw        x16, 16(x0)
-
-  /* 65537 = 0b10000000000000001
-               ^ sqr + mult
-    out = montmul(out,out)       */
-  lw        x19, 28(x0)
-  lw        x20, 28(x0)
-  lw        x21, 28(x0)
-  jal       x1, montmul
-  /* Store result in dmem starting at dmem[dptr_c] */
-  loop      x30, 2
-    bn.sid    x8, 0(x21++)
-    addi      x8, x8, 1
-
-  /* out = montmul(in,out)       */
-  lw        x19, 20(x0)
-  lw        x20, 28(x0)
-  lw        x20, 28(x0)
   jal       x1, montmul
-
-  /* store multiplication result in output buffer */
-  lw        x21, 28(x0)
-  li        x8, 4
+  /* store result in dmem starting at dmem[dptr_out] */
   loop      x30, 2
-    /* store selected limb to dmem */
     bn.sid    x8, 0(x21++)
     addi      x8, x8, 1
 
-  /* 65537 = 0b10000000000000001
-                ^<< 16 x sqr >>^   */
-  loopi      16, 8
-    /* square: out = montmul(out, out) */
+  /* 16 consecutive Montgomery squares on the outbut buffer, i.e. after loop:
+     out_buf <= out_buf^65536*R mod M */
+  loop      x29, 8
+    /* out_buf  = *x28 = *dmem[28]
+               <= montmul(*x28, *x20) = montmul(*dptr_out, *dptr_out)
+                = out_buf^2*R mod M */
     lw        x19, 28(x0)
     lw        x20, 28(x0)
     lw        x21, 28(x0)
     jal       x1, montmul
-    /* Store result in dmem starting at dmem[dptr_c] */
+    /* Store result in dmem starting at dmem[dptr_out] */
     loop      x30, 2
       bn.sid    x8, 0(x21++)
       addi      x8, x8, 1
     nop
 
-  /* 65537 = 0b10000000000000001
-                          mult ^
-     out = montmul(in,out)       */
+  /* final multiplication and convertion of result from Montgomery domain
+     out_buf  = *x28 = *dmem[28]
+             <= montmul(*x28, *x20) = montmul(*dptr_sig, *dptr_out)
+              = out_buf*sig/R mod M = sig^65537 mod M */
   lw        x19, 20(x0)
   lw        x20, 28(x0)
   lw        x21, 28(x0)
   jal       x1, montmul
 
-  /* store multiplication result in output buffer */
-  lw        x21, 28(x0)
+  /* Final conditional subtraction of modulus if mod >= out_buf. This could
+     be done in variable time, but for the sake of reduced code we use a loop
+     with N cycles. */
+  bn.add    w31, w31, w31
+  li        x17, 16
+  loop      x30, 4
+    bn.movr   x11, x8++
+    bn.lid    x9, 0(x16++)
+    bn.subb   w2, w2, w3
+    bn.movr   x17++, x11
+  csrrs     x2, 0x7c0, x0
+  /* Fixme: currently we subtract the modulus if out_buf == M. This should
+            never happen in an RSA context. We could catch this and raise an
+            alert. */
```

Comment:
Good point about raising an alert, I opened https://github.com/lowRISC/opentitan/issues/6147 to discuss and document the best way of doing so.